homeassistant/components/octoprint/config_flow.py

```diff
@@ -0,0 +1,198 @@
+        data[CONF_HOST], session, data[CONF_PORT], data[CONF_SSL], data[CONF_PATH]
+    )
+    octoprint.set_api_key(data[CONF_API_KEY])
+
+    try:
+        discovery = await octoprint.get_discovery_info()
+    except ApiError as conn_err:
+        _LOGGER.error("Error setting up OctoPrint API: %r", conn_err)
+        raise CannotConnect from conn_err
+
+    # Return info that you want to store in the config entry.
+    return {"title": data[CONF_HOST], "uuid": discovery.upnp_uuid}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for OctoPrint."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+    api_key_task = None
+    user_input: dict = None
+
+    def __init__(self):
+        """Handle a config flow for OctoPrint."""
+        self.discovery_schema = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        if user_input is None:
+            data = self.discovery_schema or _schema_with_defaults()
+            return self.async_show_form(step_id="user", data_schema=data)
+
+        self.user_input = user_input
+        if CONF_API_KEY in user_input:
+            return await self.async_step_finish(user_input)
+
+        self.api_key_task = None
+        return await self.async_step_get_api_key(user_input)
+
+    async def async_step_get_api_key(self, user_input=None):
+        """Get an Application Api Key."""
+        if not self.api_key_task:
+            self.api_key_task = self.hass.async_create_task(self._async_get_auth_key())
+            return self.async_show_progress(
+                step_id="get_api_key", progress_action="get_api_key"
+            )
+
+        try:
+            await self.api_key_task
+        except ApiError as err:
+            _LOGGER.error("Failed to get an application key : %s", err)
+            return self.async_show_progress_done(next_step_id="auth_failed")
+
+        return self.async_show_progress_done(next_step_id="finish")
+
+    async def async_step_finish(self, user_input=None):
+        """Finish the configuration setup."""
+        user_input = user_input or self.user_input
+        error = None
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except CannotConnect:
+            error = "cannot_connect"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            error = "unknown"
+        else:
+            await self.async_set_unique_id(info["uuid"], raise_on_progress=False)
+            self._abort_if_unique_id_configured()
+            return self.async_create_entry(title=info["title"], data=user_input)
+
+        return self._create_setup_failure(user_input, error)
+
+    async def async_step_auth_failed(self, user_input):
+        """Handle api fetch failure."""
+        return self._create_setup_failure(self.user_input, "auth_failed")
```

Comment:
Move back to the `async_step_get_api_key` instead without user input passed. Add a check in that step for user input is None and show an error form in that case.

Reply:
the get_api_key step doesn't have a form attached to it, should I call async_step_user() with the error set to at property of the configflow?

Reply:
I suggest to add a form to `async_step_get_api_key`.

Reply:
I may have a misunderstanding of the configflow steps, so forgive me if I do. The form that async_step_get_api_key would show would be identical to async_step_user. It seems like if async_step_get_api_key, the appropriate action would be to move back to the user step, show the user an error, and allow them to retry when they are ready. 
Is the correct thing to reproduce the user form in this step vs moving back to the user step?

Reply:
Errors are specific to the step that created them. We can't pass errors from one step to another step.

If we would remove the api key step and just make that a helper instead we could have it raise an exception that we catch in the user step and show an error form. But then we can't use the progress action like we do now.

Reply:
Maybe move the form from the user step and just have it in the api key step?